Add clarifications for SSL_CTX_set1_curves

andrewkdinh opened the pull request:
Adds some more examples to SSL_CTX_set1_curves

Fixes https://github.com/openssl/openssl/issues/26883

- [x] documentation is added or updated

vdukhovni commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -174,6 +174,31 @@ and the client supports C<P-384> including key share for this group, a HRR will
+- "C:*F": The server prefers "C" from the second group tuple "B:C", so an HRR is
+  triggered for "C".
+
+- "C:*D": The server prefers "C" even though a key share for "D" has already
+  been sent, so an HRR is triggered for "C".
```
This is false. Within a tuple existing keyshares trump preference order.

kroeckx commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -174,6 +174,31 @@ and the client supports C<P-384> including key share for this group, a HRR will
+- "B:*C": The server prefers "B" from the first group tuple "A:B", so an HRR is
+  triggered for "B".
+
+- "C:*F": The server prefers "C" from the second group tuple "B:C", so an HRR is
```
The second group is "C:D".

kroeckx commented:
So I think all the examples are with client preference, but we also support server preference. Maybe it would be useful to have examples of that too.

vdukhovni commented:
> So I think all the examples are with client preference, but we also support server preference. Maybe it would be useful to have examples of that too.

I thought the original text covered this adequately, but the difference is relatively minor, it affects which of the elements of the most preferred tuple that over overlaps with the client's supported groups is chosen in HRR (the one most preferred by the client or the server) or else (not HRR) which keyshare is chosen for the server hello when multiple keyshares are sent (the one most preferred by the client or by the server).

Tuples determine whether HRR does or does not happen, and order within tuples (or client supported group list) drives server and client preference.

t8m commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -174,6 +174,33 @@ and the client supports C<P-384> including key share for this group, a HRR will
 triggered for C<P-521> despite the availability of a key share for P-384, which
 overlaps with a lower priority server-side tuple.
 
+For example, consider the same server but different client with I<list>
```
Please consider changing the first sentence on line 115 to:
`Group tuples of comparable security are defined by separating them from each other by a tuple separator C</>.`

mattcaswell replied:
Our documentation normally has a separate `=HEAD1 EXAMPLES` section specifically for examples

mattcaswell commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -174,6 +174,33 @@ and the client supports C<P-384> including key share for this group, a HRR will
+
+- "*C:*D": The server accepts the "C" key share.
+
+- "F": The server accepts "F" key share.
```
Shouldn't we be explicit in these examples about whether we are using client preference or server preference - and include some examples of how things might turn out differently depending on which one you choose.

andrewkdinh replied:
I've moved the examples to its own section and specified differences between client and server preference. Please help me check that all the examples are correct

mattcaswell commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,67 @@ returns -1.
+- "*C:*D": The server accepts the "C" key share.
+
+- "F": The server will trigger an HRR for "F" because it did not receive a key
+share for "F".
```
This client list is an example where no keyshare `*` is present at all. In this case this rule applies:

> If no C<*> is used anywhere in the list, a single key share for the leftmost valid group is sent.

So "F" is actually equivalent to "*F". The server will accept the key share and no HRR occurs.

mattcaswell commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,67 @@ returns -1.
+  - Server preference: The server uses the key share for "C".
+
+- "*E:C:A"
+  - Client preference: The server uses the key share for "E".
```
This seems wrong? Don't we get an HRR for A in this case (i.e. same as for server preference)?

mattcaswell commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,67 @@ returns -1.
+for "A", so an HRR is triggered for "A".
+
+- "E:*C:A"
+  - Client preference: The server uses the key share for "C". Within a tuple,
```
Again I think we get an HRR for A in this case

mattcaswell commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,67 @@ returns -1.
+
+- "*E:B:*A"
+  - Client preference: The server uses the key share for "E".
+  - Server preference: The server uses the key share for "A".
```
Both will use the key share for "A"?

mattcaswell commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,67 @@ returns -1.
+I<list> is "P-521:*P-384" when connecting to such a server, meaning that the
+client supports C<P-521> but does not send a key share for this group to the
+server, and the client supports C<P-384> including key share for this group.
+With server preference a HRR will be triggered for C<P-521> despite the
```
Doesn't this apply to both client and server preference?

andrewkdinh commented:
@mattcaswell Ah ok, I think I understand it now. With client preference, essentially it's still still up to what the server prefers, but instead we only consider the client preference if there are multiple options of comparable security. I've updated the documentation and added some more examples.

vdukhovni commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,63 @@ returns -1.
+despite the availability of a key share for P-384, which overlaps with a lower
+priority server-side tuple.
+
+Consider the same server and server preference but different client with I<list>
+"P-521:P-384:*ffdhe2048". When the client attempts connecting to the server,
+only a keyshare for ffdhe2048 will be sent. But because P-521 and P-384 are
+supported by the client, an HRR will be triggered for P-521 (preferred by the
+server).
+
```
To me, this doesn't look materially different from the previous example. Perhaps drop it?

vdukhovni commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,63 @@ returns -1.
+"E:F". Here are some examples of a client I<list> where setting server/client
+preference will not change the outcome:
+
+- "A:D:*F": Both prefer "A", but the server did not receive a key share for "A",
+so an HRR is triggered for "A".
```
The result is of course correct, but the reason isn't that the server didn't receive a keyshare for "A", but rather that the server didn't receive a keyshare for the most-preferred tuple in which there's at least one group supported by both.

The invariant is that the selected keyshare is always from the most preferred tuple in which there's at least one group in common. If there are one or more keyshares for that group, one of them is chosen based on server/client preference. If there are no keyshares sent for that group an HRR will be sent for some group in that (most preferred with overlap) tuple.
Unless of course there are simply no groups in common...

vdukhovni commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,63 @@ returns -1.
+- "F": Even though it is not prepended with a "*", the client will send a key
+share for "F". The server will then accept the key share for "F".
+
+- "*E:C:A": The server prefers "A" but did not receive a key share
+for "A", so an HRR is triggered for "A".
```
Again, what the server strongly prefers is the "A:B" tuple, the secondary preference for "A" rather than
"B" is only relevant when keyshares are present for both, or for neither, but both are supported by the client, and server preference is in effect.

vdukhovni commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,63 @@ returns -1.
+
+- "B:A:*C"
+  - Client preference: The server triggers an HRR for "B". For the server,
+"A" and "B" are considered comparable in security to each other. But because the
```
```suggestion
"A" and "B" are considered comparable in security. But because the
```

vdukhovni commented on doc/man3/SSL_CTX_set1_curves.pod:
```diff
@@ -263,6 +257,58 @@ returns -1.
 
+=head1 EXAMPLES
+
+Assume the server I<list> is "P-521:P-256/P-384/X25519:ffdhe2048" and client
+I<list> is "P-521:*P-384" when connecting to such a server, meaning that the
+client supports C<P-521> but does not send a key share for this group to the
+server, and the client supports C<P-384> including key share for this group.
+With both server and client preference, an HRR will be triggered for C<P-521>
+despite the availability of a key share for P-384, which overlaps with a lower
+priority server-side tuple.
```
The examples are not substantially better. No major issues. I just feel that with the examples below, that really cover the issues well, the above "concrete" example isn't nearly as good, and the presence of specific groups that may or may not be the ones used in some scenario is more of a distraction, so I'd drop this one too...  Perhaps others will have a different view...

vdukhovni replied:
[ Edited from a  typo that had "not" instead of "now". ]

openssl-machine commented:
This pull request is ready to merge

t8m commented:
Partially squashed and merged to the master branch. Thank you.
